connectivity/tests/encryption.go

```diff
@@ -54,56 +74,134 @@ func getInterNodeIface(ctx context.Context, t *check.Test, clientHost *check.Pod
+
+	if enc, ok := t.Context().Feature(features.EncryptionPod); wgEncap && tunnelEnabled && ok &&
+		enc.Enabled && enc.Mode == "wireguard" {
+
+		// Captures the following:
+		// - Any VXLAN/Geneve pkt client host <-> server host. Such a pkt might
+		//   contain a leaked pod-to-pod unencrypted pkt. We could have made this
+		//   filter more fine-grained (i.e., to filter an inner pkt), but that
+		//   requires nasty filtering based on UDP offsets (pcap-filter doesn't
+		//   filtering of VXLAN/Geneve inner layers).
+		// - Any pkt client <-> server with a given proto. This might be useful
+		//   to catch any regression in the DP which makes the pkt to bypass
+		//   the VXLAN tunnel.
+		//
+		// Some explanations:
+		// - "udp[8:2] = 0x0800" compares the first two bytes of an UDP payload
+		//   against VXLAN commonly used flags. In addition we check against
+		//   the default Cilium's VXLAN port (8472).
+		// - To catch Geneve traffic we cannot use the "geneve" filter, as it shifts
+		//   offset of a filted packet which invalidates the later part of the
+		//   filter. Thus this poor UDP/6081 check.
+		tunnelFilter := "(udp and (udp[8:2] = 0x0800 or dst port 8472 or dst 6081))"
+		filter := fmt.Sprintf("(%s and host %s and host %s) or (host %s and host %s and %s)",
+			tunnelFilter,
+			clientHost.Address(features.IPFamilyV4), serverHost.Address(features.IPFamilyV4),
+			client.Address(ipFam), server.Address(ipFam), protoFilter)
+
+		return filter
+
 	}
 
-	srcIPStr := strings.TrimRight(srcIP.String(), "\n\r")
-	if srcIPStr != client.Address(ipFam) {
-		filter = fmt.Sprintf("( %s or src host %s )", filter, srcIPStr)
+	filter := fmt.Sprintf("src host %s", client.Address(ipFam))
+	dstIP := server.Address(ipFam)
+
+	if tunnelStatus, ok := t.Context().Feature(features.Tunnel); ok && tunnelStatus.Enabled {
+		cmd := []string{
+			"/bin/sh", "-c",
+			fmt.Sprintf("ip -o route get %s | grep -oE 'src [^ ]*' | cut -d' ' -f2",
+				dstIP),
+		}
+		t.Debugf("Running %s", strings.Join(cmd, " "))
+		srcIP, err := clientHost.K8sClient.ExecInPod(ctx, clientHost.Pod.Namespace,
+			clientHost.Pod.Name, "", cmd)
+		if err != nil {
+			t.Fatalf("Failed to get IP route: %s", err)
+		}
+
+		srcIPStr := strings.TrimRight(srcIP.String(), "\n\r")
+		if srcIPStr != client.Address(ipFam) {
+			filter = fmt.Sprintf("( %s or src host %s )", filter, srcIPStr)
+		}
 	}
+
+	filter = fmt.Sprintf("%s and %s", filter, protoFilter)
+
+	// Unfortunately, we cannot use "host %s and host %s" filter here,
+	// as IPsec recirculates replies to the iface netdev, which would
+	// make tcpdump to capture the pkts (false positive).
+	filter = fmt.Sprintf("%s and dst host %s", filter, dstIP)
+
 	return filter
 }
 
+// isWgEncap checks whether packets are encapsulated before encrypting with WG.
+//
+// In v1.14, it's an opt-in, and controlled by --wireguard-encapsulate.
+// TODO(brb; after merging v1.14, add vsn check): In v1.15, it's mandatory enabled.
```

Comment:
Should this TODO be addressed?

Reply:
I will address this in a separate PR (the related code hasn't been merged to Cilium's `main`).